File: DesignPatterns/01_GOF23/DP_examples/Python/02_Structural/Proxy/Proxy.py
from Subject import Image

# ImageProxy 是图像代理类，实现了 Image 抽象基类
class ImageProxy(Image):
    def __init__(self):
        self.images = []
    
    def operation(self):
        for image in self.images:
            image.operation()
    
    def load_from_disk(self):
        for image in self.images:
            image.load_from_disk()
    
    def add_image(self, image):
        self.images.append(image)
    
    def add_images(self, *images):
        self.images.extend(images)
    
    def remove(self, image):
        if image in self.images:
            self.images.remove(image)
            return True
        return False

File: DesignPatterns/01_GOF23/DP_examples/CSharp/02_Structural/Proxy/Proxy.cs
using System.Linq;

namespace Structural.Proxy;

public class ImageProxy : Image
{
    public List<Image> Images = new();
    public void Operation()
    {
        foreach (var image in Images)
            image.Operation();
    }
    public void loadFromDisk()
    {
        foreach (var item in Images)
            item.loadFromDisk();
    }
    public void AddImage(Image image)
    {
        Images.Add(image);
    }
    public void AddImages(params IEnumerable<Image> imgs) {
        Images.AddRange(imgs);
    }

    public bool Remove(Image image)
    {
        return Images.Remove(image);
    }
}